File: BST/pathSum.js
// does a BST path from root to leaf equal targetSum

var hasPathSum = function (root, targetSum) {
  if (!root) return false;

  const nodeStack = [];
  nodeStack.push(root);
  while (nodeStack.length >= 1) {
    let currentNode = nodeStack.pop();

    if (currentNode.left !== null) {
      currentNode.left.val += currentNode.val;
      nodeStack.push(currentNode.left);
    }

    if (currentNode.right !== null) {
      currentNode.right.val += currentNode.val;
      nodeStack.push(currentNode.right);
    }

    // if leaf node
    if (
      currentNode.left === null &&
      currentNode.right === null &&
      currentNode.val === targetSum
    ) {
      return true;
    }
  }
  return false;
};

//recursive version dfs
var hasPathSum = function (root, targetSum) {
  return dfs(root, 0, targetSum);
};

function dfs(node, currentSum, targetSum) {
  if (!node) return false;
  currentSum += node.val;

  //leaf node
  if (node.left === null && node.right === null) {
    return currentSum === targetSum;
  }

  return (
    dfs(node.left, currentSum, targetSum) ||
    dfs(node.right, currentSum, targetSum)
  );
}

File: BST/path_sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def hasPathSum(self, root: TreeNode, targetSum: int) -> bool:
        if root is None:
            return False
        
        node_stack = [root]
        while len(node_stack) >= 1: 
            current_node = node_stack.pop()
            
            if current_node.left:
                current_node.left.val += current_node.val
                node_stack.append(current_node.left)
            
            if current_node.right: 
                current_node.right.val += current_node.val
                node_stack.append(current_node.right)
            
            if current_node.left == None and current_node.right == None: 
                if current_node.val == targetSum: 
                    return True
        return False